File: sessions_mount_dir/djx6bg6ycfbcv2w07yftbf0uc/backend/app/routers/booking.py
from datetime import datetime
from typing import List
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from ..services.booking_service import BookingService
from ..models import Booking
from ..core.security import get_current_user
from ..core.monitoring import log_booking_attempt
from ..dependencies import get_db, get_redis

router = APIRouter(prefix="/bookings", tags=["bookings"])

@router.post("/", response_model=dict)
def create_booking(
    course_id: int,
    booking_time: datetime,
    db: Session = Depends(get_db),
    redis_client = Depends(get_redis),
    current_user = Depends(get_current_user)
):
    """
    Create a new booking for a course.
    """
    booking_service = BookingService(db, redis_client)
    
    # Log booking attempt for monitoring
    log_booking_attempt(current_user.id, course_id)
    
    try:
        booking = booking_service.create_booking(
            user_id=current_user.id,
            course_id=course_id,
            booking_time=booking_time
        )
        return {
            "message": "Booking created successfully",
            "booking_id": booking.id,
            "course_id": booking.course_id,
            "user_id": booking.user_id,
            "booking_time": booking.booking_time.isoformat()
        }
    except HTTPException as e:
        raise e
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="An unexpected error occurred while creating the booking"
        )

@router.get("/", response_model=List[dict])
def get_user_bookings(
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """
    Get all bookings for the current user.
    """
    bookings = db.query(Booking).filter(Booking.user_id == current_user.id).all()
    return [
        {
            "id": booking.id,
            "course_id": booking.course_id,
            "booking_time": booking.booking_time.isoformat(),
            "status": booking.status
        }
        for booking in bookings
    ]

@router.get("/{booking_id}", response_model=dict)
def get_booking(
    booking_id: int,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """
    Get a specific booking by ID.
    """
    booking = db.query(Booking).filter(
        Booking.id == booking_id,
        Booking.user_id == current_user.id
    ).first()
    
    if not booking:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Booking not found"
        )
    
    return {
        "id": booking.id,
        "course_id": booking.course_id,
        "booking_time": booking.booking_time.isoformat(),
        "status": booking.status
    }

@router.delete("/{booking_id}", response_model=dict)
def cancel_booking(
    booking_id: int,
    db: Session = Depends(get_db),
    redis_client = Depends(get_redis),
    current_user = Depends(get_current_user)
):
    """
    Cancel a booking.
    """
    booking_service = BookingService(db, redis_client)
    
    try:
        success = booking_service.cancel_booking(booking_id, current_user.id)
        if not success:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Booking not found or cannot be cancelled"
            )
        return {"message": "Booking cancelled successfully"}
    except HTTPException as e:
        raise e
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="An unexpected error occurred while cancelling the booking"
        )

File: sessions_mount_dir/cnj86j24me1u82atid76xx9gc/frontend/src/api/booking.js
import axios from 'axios';

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000/api/v1';

/**
 * 获取所有预约记录
 * @returns {Promise} 预约记录列表
 */
export const getAllBookings = async () => {
  try {
    const response = await axios.get(`${API_BASE_URL}/bookings`);
    return response.data;
  } catch (error) {
    console.error('Error fetching bookings:', error);
    throw error;
  }
};

/**
 * 根据用户ID获取预约记录
 * @param {string} userId - 用户ID
 * @returns {Promise} 用户的预约记录列表
 */
export const getBookingsByUser = async (userId) => {
  try {
    const response = await axios.get(`${API_BASE_URL}/bookings/user/${userId}`);
    return response.data;
  } catch (error) {
    console.error('Error fetching user bookings:', error);
    throw error;
  }
};

/**
 * 创建新预约
 * @param {Object} bookingData - 预约数据
 * @param {string} bookingData.course_id - 课程ID
 * @param {string} bookingData.user_id - 用户ID
 * @returns {Promise} 创建的预约记录
 */
export const createBooking = async (bookingData) => {
  try {
    const response = await axios.post(`${API_BASE_URL}/bookings`, bookingData);
    return response.data;
  } catch (error) {
    console.error('Error creating booking:', error);
    throw error;
  }
};

/**
 * 取消预约
 * @param {string} bookingId - 预约ID
 * @returns {Promise} 取消结果
 */
export const cancelBooking = async (bookingId) => {
  try {
    const response = await axios.delete(`${API_BASE_URL}/bookings/${bookingId}`);
    return response.data;
  } catch (error) {
    console.error('Error cancelling booking:', error);
    throw error;
  }
};

/**
 * 获取课程预约状态
 * @param {string} courseId - 课程ID
 * @returns {Promise} 课程预约状态信息
 */
export const getCourseBookingStatus = async (courseId) => {
  try {
    const response = await axios.get(`${API_BASE_URL}/courses/${courseId}/booking-status`);
    return response.data;
  } catch (error) {
    console.error('Error fetching course booking status:', error);
    throw error;
  }
};

/**
 * 检查用户是否已预约某课程
 * @param {string} courseId - 课程ID
 * @param {string} userId - 用户ID
 * @returns {Promise} 预约状态
 */
export const checkUserBooking = async (courseId, userId) => {
  try {
    const response = await axios.get(`${API_BASE_URL}/bookings/check`, {
      params: { course_id: courseId, user_id: userId }
    });
    return response.data;
  } catch (error) {
    console.error('Error checking user booking:', error);
    throw error;
  }
};
